### ISSUE TITLE
Spyk mocked property not used inside class

### ISSUE THREAD

**Issue by davidhiendl** [mockk/mockk#279]
When mocking a property via spky (see example below) any external access to this property works just fine.
However when the property is accessed inside the class it is not mocked.
This might relate to #263 (not sure)

I've attached a reproducible example below.

### Expected Behavior
Mocked properties should be overridden wherever they are accessed.

### Current Behavior
The property access is only working outside the class.

#### Steps to Reproduce
Run the supplied sample test.

#### Context
* MockK: since at least 1.9.2, re-tested with 1.9.3 (same issue)
* OS: Linux
* Kotlin: 1.3.21
* JDK version: jdk-1.8
* JUnit version: 1.3.2 (io.kotlintest:kotlintest-runner-junit5:3.2.1)
* Type of test: unit test OR android instrumented test

#### Stack trace
```
java.lang.IllegalStateException: in-class access is false
	at DummyClass.useInternal(MockPropertyTest.kt:10)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.mockk.proxy.jvm.advice.MethodCall.call(MethodCall.kt:14)
	at io.mockk.proxy.jvm.advice.SelfCallEliminatorCallable.call(SelfCallEliminatorCallable.kt:14)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper.handleOriginalCall(JvmMockFactoryHelper.kt:83)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper.access$handleOriginalCall(JvmMockFactoryHelper.kt:20)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1$invocation$$inlined$stdFunctions$lambda$1.invoke(JvmMockFactoryHelper.kt:28)
	at io.mockk.impl.stub.MockKStub$handleInvocation$originalPlusToString$1.invoke(MockKStub.kt:227)
	at io.mockk.impl.stub.SpyKStub.defaultAnswer(SpyKStub.kt:15)
	at io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:42)
	at io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:53)
	at io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:263)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:25)
	at io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:20)
	at DummyClass.useInternal(MockPropertyTest.kt:11)
	at MockPropertyTest.Mock Test Property(MockPropertyTest.kt:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at kotlin.reflect.jvm.internal.calls.CallerImpl$Method.callMethod(CallerImpl.kt:97)
	at kotlin.reflect.jvm.internal.calls.CallerImpl$Method$Instance.call(CallerImpl.kt:113)
	at kotlin.reflect.jvm.internal.KCallableImpl.call(KCallableImpl.kt:106)
	at kotlin.reflect.full.KCallables.callSuspend(KCallables.kt:51)
	at io.kotlintest.specs.AbstractAnnotationSpec$createTestCase$1.invokeSuspend(AbstractAnnotationSpec.kt:82)
	at io.kotlintest.specs.AbstractAnnotationSpec$createTestCase$1.invoke(AbstractAnnotationSpec.kt)
	at io.kotlintest.runner.jvm.TestCaseExecutor$executeTestIfActive$supervisorJob$1$invokeSuspend$$inlined$map$lambda$1.invokeSuspend(TestCaseExecutor.kt:117)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

#### Minimal reproducible code

```
test {
    useJUnitPlatform()
}

dependencies {
    testImplementation 'io.kotlintest:kotlintest-runner-junit5:3.2.1'
    testImplementation "org.jetbrains.kotlin:kotlin-test:${kotlinVersion}"
    testImplementation "io.mockk:mockk:1.9.2"
}
```

```kotlin
import io.kotlintest.specs.AnnotationSpec
import io.mockk.every
import io.mockk.spyk
import kotlin.test.assertEquals

class DummyClass {
    var test = false
        internal set


    fun useInternal() : Boolean {
        if(!test) throw IllegalStateException("in-class access is false")
        return true
    }

}

class MockPropertyTest : AnnotationSpec() {

    lateinit var dummy: DummyClass

    @BeforeClass
    fun `Prepare Dummy`() {
        dummy = spyk(DummyClass())
        every { dummy.test } returns true
    }

    @Test
    fun `Mock Test Property`() {
        assertEquals(dummy.test, true) // Works
        assert(dummy.useInternal()) // Throws exception
    }

}
```

**Comment by oleksiyp** [comment]
Thanks

**Comment by davidhiendl** [comment]
@oleksiyp Is there something I missed with the report? Is this the intended behavior?

**Comment by oleksiyp** [comment]
No everything is fine. Good report. Pretty weird error message although.

**Comment by davidhiendl** [comment]
The error is artificial, I just picked the text to signal whats going wrong. If the mock would work correctly then no exception would be thrown. The example constructed from a real-world test where I was mocking internal state properties in order to skip certain internal actions.

My suspicions as to why this is happening is that the Kotlin compiler optimizes access to properties inside the class and skips calls to getter's when they are not needed. This believe is based on the following modification to the `DummyClass` class.

By introducing a side-effect on the accessor the compiler can no longer optimize the call as it cannot remove the side-effect therefor 'fixing' the test (but that is not a solution obviously).

```kotlin
class DummyClass {
    var test = false
        get() {
            println(field)
            return field
        }

    fun useInternal(): Boolean {
        if (!test) throw IllegalStateException("in-class access failed")
        return true
    }
}
```

**Comment by oleksiyp** [comment]
I see. This optimization issue is most probable. Nothing can be done about it

**Comment by davidhiendl** [comment]
Is it possible to prevent the compiler from generating these specific optimizations (just for test compile scope, eg. is there a flag)?
Otherwise this issue should maybe be raised with Kotlin itself?

My current workaround is to use the internal modifier instead of private for properties that are changed during tests and instead of using `every { ... } ...` I just set them to the desired value. This keeps properties isolated between modules but is kinda ugly and I think is more error prone. Thankfully I only have a few such cases.

**Comment by oleksiyp** [comment]
Keeping it closed in favour of #104
